File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return checkBST(root);
    }
private:
    bool checkBST(TreeNode *root, TreeNode *min = nullptr, TreeNode *max = nullptr) {
        if (!root) return true;

        if (min && root->val <= min->val) return false;

        if (max && root->val >= max->val) return false;

        return checkBST(root->left, min, root) && checkBST(root->right, root, max);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, m):
            good = 0
            if node is None:
                return 0
            if node.val >= m:
                good += 1

            n_max = max(node.val, m)
            good += dfs(node.left, n_max)
            good += dfs(node.right, n_max)
            return good

        return dfs(root, float("-inf"))
